internal/handler/common/update.go

```diff
@@ -350,653 +351,570 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
-					docValue,
-					must.NotFail(doc.Get("_id")),
-				),
-				command,
-			)
-		default:
-			return false, err
+	case errors.Is(err, handlerparams.ErrUnexpectedLeftOpType):
+		return false, NewUpdateError(
+			handlererrors.ErrTypeMismatch,
+			fmt.Sprintf(
+				`Cannot multiply with non-numeric argument: {%s: %#v}`,
+				mulKey,
+				mulValue,
+			),
+			command,
+		)
+	case errors.Is(err, handlerparams.ErrUnexpectedRightOpType):
+		k := mulKey
+		if path.Len() > 1 {
+			k = path.Suffix()
 		}
-	}
 
-	return changed, nil
+		return false, NewUpdateError(
+			handlererrors.ErrTypeMismatch,
+			fmt.Sprintf(
+				`Cannot apply $mul to a value of non-numeric type. `+
+					`{_id: %s} has the field '%s' of non-numeric type %s`,
+				types.FormatAnyValue(must.NotFail(doc.Get("_id"))),
+				k,
+				handlerparams.AliasFromType(docValue),
+			),
+			command,
+		)
+	case errors.Is(err, handlerparams.ErrLongExceededPositive), errors.Is(err, handlerparams.ErrLongExceededNegative):
+		return false, NewUpdateError(
+			handlererrors.ErrBadValue,
+			fmt.Sprintf(
+				`Failed to apply $mul operations to current value ((NumberLong)%d) for document {_id: "%s"}`,
+				docValue,
+				must.NotFail(doc.Get("_id")),
+			),
+			command,
+		)
+	case errors.Is(err, handlerparams.ErrIntExceeded):
+		return false, NewUpdateError(
+			handlererrors.ErrBadValue,
+			fmt.Sprintf(
+				`Failed to apply $mul operations to current value ((NumberInt)%d) for document {_id: "%s"}`,
+				docValue,
+				must.NotFail(doc.Get("_id")),
+			),
+			command,
+		)
+	default:
+		return false, err
+	}
 }
 
 // processCurrentDateFieldExpression changes document according to $currentDate operator.
 // If the document was changed it returns true.
-func processCurrentDateFieldExpression(doc *types.Document, currentDateVal any) (bool, error) {
+func processCurrentDateFieldExpression(doc *types.Document, field string, value any) (bool, error) {
 	var changed bool
-	currentDateExpression := currentDateVal.(*types.Document)
 
 	now := time.Now().UTC()
-	keys := currentDateExpression.Keys()
-	sort.Strings(keys)
 
-	for _, field := range keys {
-		currentDateField := must.NotFail(currentDateExpression.Get(field))
+	var dateType any
 
-		switch currentDateField := currentDateField.(type) {
-		case *types.Document:
-			currentDateType, err := currentDateField.Get("$type")
-			if err != nil { // default is date
-				doc.Set(field, now)
-				changed = true
-				continue
-			}
-
-			currentDateType = currentDateType.(string)
-			switch currentDateType {
-			case "timestamp":
-				doc.Set(field, types.NextTimestamp(now))
-				changed = true
-
-			case "date":
-				doc.Set(field, now)
-				changed = true
-			}
+	switch value := value.(type) {
+	case *types.Document:
+		// ignore error, error cases were validated in validateCurrentDateExpression
+		dateType, _ = value.Get("$type")
+	case bool:
+		dateType = "date"
```

Comment:
The `dateType` refers to BSON type - `Date` or `timestamp`. In MongoDB, if `bool` value is passed for a specific key with `$currentDate` operator, then current date will be set as `Date` type. That is what this line is referring to.

Perhaps, `dateType` name is confusing. I will rename it `setValType` and add comments explaining the behaviour of `$currentDate` operator. Is that fine? Or you would still prefer the previous implementation? Because I changed the prev impl, because it was confusing to me with nested switch cases. :sweat_smile:

Reply:
Ah got it, you wanted to avoid nested switch case. Looks fine now :hugs: